firestore/bulkwriter.go

```diff
@@ -0,0 +1,521 @@
+// Each call to Create, Update, Set, and Delete can return a value and error.
+type BulkWriter struct {
+	database        string                      // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	reqs            int64                       // total number of requests sent
+	start           time.Time                   // when this BulkWriter was started
+	vc              *vkit.Client                // internal client
+	isOpen          bool                        // signal that the BulkWriter is closed
+	startFlush      chan struct{}               // signal the beginning of flushing operation
+	isFlushing      bool                        // determines whether we're in a flush state
+	endFlush        chan struct{}               // signal the end of flushing
+	received        chan bulkWriterJob          // incoming write requests
+	scheduled       chan bulkWriterRequestBatch // scheduled requests
+	sendingQueue    []bulkWriterJob             // queue of requests to send
+	backlogQueue    []bulkWriterJob             // queue of requests to store in memory during flush operation
+	maxOpsPerSecond int                         // number of requests that can be sent per second.
+	docUpdatePaths  sync.Map                    // document paths with corresponding writes in the queue.
+	waitTime        float64                     // time to wait in between requests; increase exponentially
+	writesLog       sync.Map                    // stores writes provided, sent, and received
+	cancel          context.CancelFunc          // context to send cancel message
+	mu              sync.Mutex                  // locks concurrent access
+}
+
+// newBulkWriter creates a new instance of the BulkWriter. This
+// version of BulkWriter is intended to be used within go routines by the
+// callers.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+
+	var sm sync.Map
+	sm.LoadOrStore(writesProvided, 0)
+	sm.LoadOrStore(writesReceived, 0)
+	sm.LoadOrStore(writesSent, 0)
+
+	var dsm sync.Map
+
+	ct, cancel := context.WithCancel(ctx)
+
+	bw := BulkWriter{
+		vc:              c.c,
+		database:        database,
+		isOpen:          true,
+		startFlush:      make(chan struct{}),
+		isFlushing:      false,
+		endFlush:        make(chan struct{}),
+		received:        make(chan bulkWriterJob),
+		scheduled:       make(chan bulkWriterRequestBatch),
+		start:           time.Now(),
+		maxOpsPerSecond: DefaultStartingMaximumOpsPerSecond,
+		waitTime:        coolingPeriodMillis,
+		writesLog:       sm,
+		cancel:          cancel,
+		docUpdatePaths:  dsm,
+	}
+
+	// Start the call receiving thread and request sending thread
+	// NOTE: We are using memory to get more speed, be aware of memory usage
+	go bw.receiver(withResourceHeader(ct, c.path()))
+	go bw.scheduler(withResourceHeader(ct, c.path()))
+
+	// TODO(telpirion): Create a retry thread? Or separate retry queue?
+
+	return &bw
```

Comment:
nit: it's more common to initialize a type with the `&` operator, instead of declaring it first, and then taking its address later.

Reply:
Done.